src/DependencyInjection/DI/src/ServiceLookup/DynamicServiceProviderEngine.cs

```diff
@@ -19,12 +19,30 @@ public DynamicServiceProviderEngine(IEnumerable<ServiceDescriptor> serviceDescri
             var callCount = 0;
             return scope =>
             {
+                // Resolve the result before we increment the call count, this ensures that singletons
+                // won't cause any side effects during the compilation of the resolve function.
+                var result = RuntimeResolver.Resolve(callSite, scope);
+
                 if (Interlocked.Increment(ref callCount) == 2)
                 {
-                    Task.Run(() => base.RealizeService(callSite));
+                    // Don't capture the ExecutionContext when forking to build the compiled version of the
+                    // resolve function
+                    ThreadPool.UnsafeQueueUserWorkItem(state =>
```

Comment:
@stephentoub is this the right way to suppress ExecutionContext?  Should this instead just use `ExecutionContext.SuppressFlow()`?

Reply:
Yah, this is best 😄

Reply:
Ah, this is why David pinged me earlier :-)

Yeah, this is good

Reply:
I see, I wasn't sure if ThreadPool would be supported in all the same cases that ExecutionContext+Task would.  I remember back in the windows phone / early UWP days we had to avoid ThreadPool.

Reply:
It's all good now. We've even been extending it, e.g. new overloads of this method.

Reply:
For the education of someone like me, what is the difference between

`ThreadPool.UnsafeQueueUserWorkItem`

vs.

`ExecutionContext.SuppressFlow + Task.Run`?

When would I use one over the other?

Reply:
> When would I use one over the other?

Basically, for fire-and-forget, use the former; if you want to later join with the operation, use the latter.